Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,488 @@
+  rw [mem_upperBounds] at hF'
+  apply Eventually.mp (expterm_eventually_le_one C n hn) (Eventually.filter_mono inf_le_right _)
+  rw [comap_principal]
+  apply eventually_of_mem (mem_principal_self _) _
+  intros z hset hexp
+  specialize hF' (abs (F f z))
+  -- Showing abs (F f z) ∈ abs ∘ F f '' closedStrip 0 1
+  specialize hF' _
+  · simp only [image_congr, mem_image, comp_apply]; use z; refine ⟨hset , refl _⟩
+  -- Combining hexp with hF' by trans
+  · exact le_trans (mul_le_mul hF' (le_refl _) (map_nonneg _ _)
+      (le_trans (map_nonneg _ _) hF')) hexp
+
+-- Proof that the edges are bounded by one
+lemma F_edge_le_one (f : ℂ → ℂ) (hB : BddAbove ((abs ∘ f) '' (closedStrip 0 1)))
+    (hz : z ∈ re ⁻¹' {0, 1}) : abs (F f z) ≤ 1 := by
+  simp only [F, Algebra.id.smul_eq_mul, map_mul]
+  cases' lt_or_eq_of_le (sSupAbsIm_nonneg f 0) with hpos hzero
+  · cases' lt_or_eq_of_le (sSupAbsIm_nonneg f 1) with h1pos h1zero
+    -- Positive case
+    · simp only [inv_interp_strip_of_pos_def z f hpos h1pos, map_mul, abs_cpow_eq_rpow_re_of_pos,
+        hpos, h1pos, sub_re, one_re, neg_re]
+      cases' hz with hz0 hz1
+      -- z.re = 0
+      · simp only [hz0, mul_one, zero_sub, Real.rpow_zero, neg_zero,
+        Real.rpow_neg_one, mul_inv_le_iff hpos]
+        rw [← hz0]
+        apply f_le_Sup f z _ hB
+        simp only [closedStrip, mem_preimage, zero_le_one, left_mem_Icc, hz0]
+      -- z.re = 1
+      · rw [mem_singleton_iff] at hz1
+        simp only [hz1, one_mul, Real.rpow_zero, sub_self, Real.rpow_neg_one,
+        mul_inv_le_iff h1pos, mul_one]
+        rw [← hz1]
+        apply f_le_Sup f z _ hB
+        simp only [closedStrip, mem_preimage, zero_le_one, hz1, right_mem_Icc]
+
+    -- Handling cases where sSupAbsIm f 0 = 0 or sSupAbsIm f 1 = 0.
+    · rw [inv_interp_strip_of_zero_def z f]
+      simp only [zero_le_one, mul_zero, map_zero]
+      right; simp only [h1zero, eq_self_iff_true]
+  · rw [inv_interp_strip_of_zero_def z f]
+    simp only [zero_le_one, mul_zero, map_zero]
+    left; rw [eq_comm]; exact hzero
+
+
+-- Edges of F' are constructed to be ≤ 1
+lemma edges_le_one (f : ℂ → ℂ) (n : ℕ) (hB : BddAbove ((abs ∘ f) '' (closedStrip 0 1)))
+    (hz : z ∈ re ⁻¹' {0, 1}) : (abs ∘ F' n f) z ≤ 1 := by
+  -- Small useful lemma
+  have hdivnat : 0 ≤ ((n : ℂ)⁻¹).re := by
+    simp only [← ofReal_nat_cast n, ← ofReal_inv n, ofReal_re, inv_nonneg, Nat.cast_nonneg]
+
+  -- Expterm ≤ 1
+  have hexp : abs (exp ((z ^ 2 - 1) * (↑n)⁻¹)) ≤ 1 := by
+    rw [abs_exp, ← re_add_im z]
+    simp only [tsub_zero, sub_re, one_re, add_im, add_zero, mul_one, mul_re,
+      zero_div, zero_mul, ofReal_re, add_re, one_im, nat_cast_im, ofReal_im, I_im,
+      zero_add, inv_im, sq, sub_im, I_re, Real.exp_le_one_iff, mul_im, mul_zero, neg_zero]
+    cases' hz with hz0 hz1
+    · simp only [hz0]
+      simp only [ofReal_zero, cpow_two, nat_cast_re, sq]
+      simp only [zero_add, mul_re, ofReal_re, I_re, mul_zero, ofReal_im, I_im, mul_one, sub_self,
```

Comment:
These consecutive simp calls should normally be combined into a single one unless it leads to some performance issue.

Reply:
Done